### ISSUE TITLE
I always forget to set post language

### ISSUE THREAD

**Issue by davidak** [mastodon/mastodon#19893]
### Steps to reproduce the problem

0. have set a default language
1. post something that is not in the default language
2. notice it after it's posted and it already has favs and replys, so i can't just redraft

### Expected behaviour

designe that prevents this issue

### Actual behaviour

i just post in my set default language

### Detailed description

one idea i have to fix it is not to have a default language and have to set it every time, but that would be very annoying

a good solution would be an algorithm to detect the language and maybe set it automatically or ask to change it. it could have an option "always set the detected languge"

translation software can detect the language pretty reliable

### Specifications

Mastodon v3.5.3 WebUI

**Comment by ashemedai** [comment]
A possible solution could be to leverage language detection from something like [chardet](https://www.npmjs.com/package/chardet).

**Comment by davidak** [comment]
They seem to have language auto-detect implemented for translating posts without specified language, so it should be easy to use it in this case as well. https://github.com/mastodon/mastodon/pull/19244

**Comment by promovicz** [comment]
As a mulilingual I would like this solved using language detection, because selecting manually it is a real nuisance.

**Comment by davidak** [comment]
Post editing helps when you notice it yourself after posting. When creating a post, it should be checked if the set language matches the detected. If not, show it e.g. my making the button red. If one often changes languages, the default language setting can provide "detect language".

**Comment by venthur** [comment]
I agree. I believe a lot of people actually forget to set the correct language when posting things. On a big instance such as mastodon.social, I limited the languages I see in the timelines to english and german and still see a firehose of toots in other languages.

I believe some language detection in the frontend, that nudges the user if the detected language does not match the selected (or ignored) language would go a long way making the timelines for everyone more enjoyable.

**Comment by blackjack75** [comment]
Not a solution for the web clients but given how many use apps this could simply be solved for a majority (?) if native clients just derived language from the onscreen keyboard used to write the post.

**Comment by venthur** [comment]
I see that mastodon is using [libretranslate](https://github.com/mastodon/mastodon/blob/30453fab80d55fc10766f0e067c31d96753ccfda/app/lib/translation_service/libre_translate.rb) which has a feature to [detect the language](https://libretranslate.com/docs/#/translate/post_detect) given some text. This could be a good starting point for a fix.

**Comment by skerit** [comment]
I'm seeing **so many** posts marked with the wrong language, the language filter feature is almost useless right now.
Most of the time English is involved: either the post is in English marked as something else, or the other way around.

I point it out to people sometimes, and the answer is always that they had no idea it was a feature or they they don't know how to change it. Doesn't help that some apps (like Tusky) don't even provide an option to change it per post.

**Comment by nemobis** [comment]
Language detection was removed in https://github.com/mastodon/mastodon/pull/17478

**Comment by das-g** [comment]
In #21631 I propose a different approach to remedy the problem of forgetting to set the post's language correctly.

**Comment by ronilaukkarinen** [comment]
I've formed a habit out of this, but I agree with everyhing above, the filtering currently doesn't work as so many forget to set the language.

**Comment by promovicz** [comment]
Some clients don't even allow setting the language, so some people give up on tagging them.

I have experimented with a Python (sic) library called Lingua for language detection on the Federated feeds. The results looked quite encouraging - especially if the detection is seeded with the users language spectrum on long posts, the language spectrum that they tag with, and possibly their client-side language selection. As such I do not understand why earlier experiments at language detection have yielded bad results. It must have been a simplistic implementation that does not seed the detector.
